[TECH] Ajouter un service d'envoi de mail via SMTP

clemlatz opened the pull request:
## :unicorn: Problème

Aujourd'hui, l'envoi de mail via SMTP est impossible et se fait uniquement via l'API de Brevo.

Cette manière de fonctionner présente plusieurs inconvénients :
- nous rendant dépendant de l’API d’un prestataire spécifique et rendrait tout changement de prestataire compliqué ;
- nous obligeant à stocker une partie de notre code (les templates d’e-mail) chez les prestataire plutôt que dans notre repository open source ;
- nous forçant à utiliser leur outil plutôt que les nôtres pour les éditer ;
- nous empêchant de tester les envois d’e-mails localement, sans connexion internet, puisque tout envoi passe nécessairement par l’API Brevo (et est donc facturable).

## :robot: Proposition

Utiliser le standard SMTP (supporté par Brevo).

Cette PR ajoute un service `smtp-mailer` qui peut être appelé à la place du classique `mailer` dans le `mail-service`. Il n'est pas encore utilisé dans cette PR. Pour l'utiliser concrètement, il faut aussi pouvoir gérer côté Pix la génération du HTML des emails à partir d'un template. Le travail pour cela a été commencé dans une autre branche (`pix-tech-use-smtp-for-account-creation-email`).

## :rainbow: Remarques

- Le dossier "smtp" a été renommé "providers" pour souligner le fait que, contrairement au niveau service, les services classiques utilisent une api HTTP et non le standard SMTP.
- Les tests des fichiers du dossier "mail" ont été déplacé pour que la structure des dossiers des tests soient équivalente à celles des dossiers contenant les implémentations.

## :100: Pour tester

### 1. Configurer l'envoi de mail via SMTP

Pour tester cette PR, il faut configurer l'envoi d'e-mail par SMTP via la variable d'environnement `MAILING_SMTP_URL`. Plus simple que d'utiliser un service d'envoi d'email tel que brevo, on peut installer un serveur smtp local de test comme [Mailpit](https://github.com/axllent/mailpit). 

Cet outil propose une interface web pour recevoir des mails ([sur le port 8025](http://localhost:8025/)) et un serveur SMTP sur le port `1025`, et donc la variable d'environnement à renseigner dans le fichier `api/.env` est :

```env
MAILING_SMTP_URL=smtp://username:password@localhost:1025/
```

La variable `MAILING_ENABLED` doit également être à `true` :

```env
MAILING_ENABLED=true
```

### 2. Utiliser `smtp-mailer` dans `mail-service`

Cette PR ajoute le service d'envoi d'e-mail via SMTP mais ne l'utilise pas encore. Pour faire un test fonctionnel, il faut donc  remplacer un appel au mailer classique par un appel à `SmtpMailer.sendEmail`, par exemple dans la méthode `sendAccountCreationEmail`:

```js
import SmtpMailer from '../../../src/shared/mail/infrastructure/services/smtp-mailer.js';
  
  // ... 
  // À la fin de la méthode sendAccountCreationEmail
  const smtpMailer = new SmtpMailer();
  return smtpMailer.sendEmail({
    from: EMAIL_ADDRESS_NO_RESPONSE,
    fromName: pixName,
    to: email,
    subject: accountCreationEmailSubject,
    html: '<h1>Welcome !</h1>',
    text: 'Welcome',
  });
```

### 3. Tester fonctionnellement

1. Créer un nouveau compte Pix
2. Dans [l'interface web de Mailpit](http://localhost:8025/), constater la réception du mail avec le bon expéditeur, destinataire, sujet, et pour corps un titre de niveau 1, "Welcome".

igorissen committed:
build(api): add nodemailer dependency

igorissen committed:
feat(api): add SmtpMailer service

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr7233.review.pix.fr)
  * [App (.org)](https://app-pr7233.review.pix.org)
  * [Orga (.fr)](https://orga-pr7233.review.pix.fr)
  * [Orga (.org)](https://orga-pr7233.review.pix.org)
  * [Certif (.fr)](https://certif-pr7233.review.pix.fr)
  * [Certif (.org)](https://certif-pr7233.review.pix.org)
  * [1D](https://1d-pr7233.review.pix.fr)
  * [Admin](https://admin-pr7233.review.pix.fr)
  * [API](https://api-pr7233.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr7233.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr7233/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr7233/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr7233/environment)

clemlatz committed:
feat(api): add MAILING_SMTP_URL config env var

clemlatz committed:
refactor(api): improve mail services directory structure

Rename "smtp" directory to "providers" to clarify that
legacy email providers do not use SMTP, but rather http APIs.

clemlatz committed:
test(api): improve mail services tests directory structure

So that test directory structure match implementation files
directory structure

dlahaye commented on api/src/shared/mail/infrastructure/services/smtp-mailer.js:
```diff
@@ -0,0 +1,29 @@
+    this.#transporter = SmtpMailer.createTransport();
+  }
+
+  static createTransport() {
+    return createTransport(config.mailing.smtpUrl);
+  }
```
ça ne serait pas plus propre de passer les vars d'env nécessaires via le constructor ?
Voir [cet article](https://notestoself.dev/posts/code-smell-env-vars-in-constructor/#:~:text=Config%20is%20a%20dependency%20like%20any%20other%2C%20and%20should%20be%20injected%20into%20the%20class%20constructor.).

clemlatz replied:
En effet, mais quelle serait l'alternative ? Si on passe `smtpUrl` en argument au constructeur, j'ai l'impression qu'on ne fait que décaler le problème, la "variable globale" qu'est la var d'env sera toujours importée, mais ailleurs. Et potentiellement à plusieurs endroits, puisqu'il faudra le faire à chaque fois qu'on instancie la classe SmtpMailer.

dlahaye replied:
Ah oui c'est parce que j'ai l'habitude d'avoir des instanciations de services dans le cadre d'injections de dépendance. Dans ce cas-là, toutes les injections se font à la racine du projet (grossomodo) et puis nul part ailleurs

clemlatz replied:
Ok, du coup une solution serait d'avoir un service "config" récupérable via l'injection de dépendance et qu'on passerait au service depuis le usecase ?

dlahaye replied:
Disons qu'à l'endroit où les injections se font, on aurait tendance à instancier le SmtpMailer en y injectant la config, puis on injecte l'instance du SmtpMailer dans le ou les usecases / services etc. qui en ont besoin.
Mais je ne sais pas si notre méthode d'injection permet cette souplesse.

dlahaye replied:
Mais en vrai ce bazar ça sert surtout quand tu as des tokens ou des url du provider pour la prod, le staging, les tests, docker etc. et que tu veux faire du e2e en décidant de quel env. tu veux utiliser

dlahaye commented on api/src/shared/mail/infrastructure/services/smtp-mailer.js:
En fouillant dans les fichiers de cette PR, je n'ai pas vu où la classe est instanciée puis utilisée 🤔

francois2metz reviewed (approved):
:rose: :hibiscus: :white_flower: :wilted_flower:  :tulip: :sunflower: :cherry_blossom:  LGTM :cherry_blossom: :sunflower:  :tulip: :wilted_flower: :white_flower: :hibiscus: :rose:
